### DISCUSSION TITLE
Async commands response order guarantees

### DISCUSSION THREAD

**Discussion by elmart-devo** [redis/lettuce#2017] [Q&A] [answered]
Hi,

I'm using the async API to send a lot of commands at once.
Then, when responses arrive, I'm adding them to a queue (not to block the netty event loop thread).
And then, in another thread, I'm processing responses in the queue.
```java
      redisAsyncCommands
        .evalsha(scriptSha, ScriptOutputType.MULTI, keys, values)
        .whenCompleteAsync((result, throwable) -> {
          responses.add(new RedisResponse(keys, values, result, throwable));
        });
```
From the docs, I expect commands to be executed in the order they were sent.
And, being a Redis standalone instance, I would have expected for responses to be delivered in order, too.
But I'm finding unordered responses in the queue.
I haven't found anything regarding response order in the docs.

So, what I need to know is:
- Is there any guarantee about response order, or do I have to manually match responses to requests if I want to process responses in order?
- What about redis cluster? does execution order guarantees still apply? what about response order here?

If somebody can shed some light on that, it would be great. Thx!

**UPDATE**: I've now seen my error above (using `whenCompleteSync` instead of `whenComplete`), after which responses are in order as I expected. But that sill leaves the questions above unanswered. Are there guarantees to this respect, both in the standalone and cluster mode? Thx.

**Comment by perlun** [comment, marked as answer]
> **UPDATE**: I've now seen my error above (using `whenCompleteSync` instead of `whenComplete`), after which responses are in order as I expected.

Good point. I assume you mean `whenCompleteAsync()` btw, which means that the code runs asynchronously => potentially in an unpredictable order

> But that still leaves the questions above unanswered. Are there guarantees to this respect, both in the standalone and cluster mode? Thx.

See this page in the documentation for more details: https://github.com/lettuce-io/lettuce-core/wiki/Command-execution-reliability#message-ordering. But simply speaking, this should be fine _as long as requests are coming from a single thread_, as described by that page.

**Reply by elmart-devo** [reply]
You're right, I meant `whenCompleteAsync`. And thx for that page! It answers my question.
